[tca9548a] Allow separate I²C bus frequencies on each multiplexer port

DAVe3283 opened the pull request:
# What does this implement/fix?

Support changing the I²C bus frequency so each port on a TCA9548A I2C multiplexer can run at a different speed.
This is for a project with I²C devices on long Ethernet cables that start to get unreliable above 50 kHz, but a I²C display locally is basically unusable at that speed.
I suspect this feature would be helpful in other situations as well.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):** N/A

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- esphome/esphome-docs#5284

## Test Environment

- [X] ESP32
- [X] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx
- [ ] nRF52840

## Example entry for `config.yaml`:

```yaml
i2c:
  id: i2c0
  frequency: 400 kHz

tca9548a:
  - i2c_id: i2c0
    channels:
      - bus_id: multiplex0channel0
        channel: 0
      - bus_id: multiplex0channel1
        channel: 1
        frequency: 100 kHz
```

## Checklist:
  - [X] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

DAVe3283 committed:
Support changing I²C bus frequency at runtime so mux ports can run at different speeds

DAVe3283 committed:
Fix debug statement

DAVe3283 committed:
Remove debug prints

DAVe3283 committed:
Update to match style guide

DAVe3283 committed:
Add test with frequency specified

DAVe3283 committed:
Remove debug comments

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

DAVe3283 committed:
Merge branch 'dev' into change-i2c-freq

DAVe3283 committed:
Improve comment

DAVe3283 committed:
Restore set_frequency function missed in merge conflict resolution

DAVe3283 commented:
There appears to be a memory leak when I repeatedly add new device configs without removing the old ones. Will fix that and test a bit.

DAVe3283 committed:
Fix a memory leak by removing previous device config

DAVe3283 committed:
Increase temporary debug log verbosity to rise above the noise

This line will go away eventually anyway

DAVe3283 committed:
Remove temporary debug message

DAVe3283 commented:
Read through the [`i2c_master_bus_add_device`](https://github.com/espressif/esp-idf/blob/2044fba6e71422446986f9ae0909b1ab67e57815/components/esp_driver_i2c/i2c_master.c#L1151) function, which allocates memory with `heap_caps_calloc` and `calloc`.
Confirmed that the [`i2c_master_bus_rm_device`](https://github.com/espressif/esp-idf/blob/2044fba6e71422446986f9ae0909b1ab67e57815/components/esp_driver_i2c/i2c_master.c#L1187) function calls `free` on everything that was created.

Removing the existing I2C device config (freeing its memory) before setting the new config/bus speed resolves the memory leak.

Graph of heap usage with the previous memory leak until just before midnight, and with the fix showing stable memory use after: 
<img width="1819" height="398" alt="plot of free heap memory" src="https://github.com/user-attachments/assets/d7c246cc-11b4-47ad-9280-76862e4caefa" />

DAVe3283 committed:
Merge remote-tracking branch 'remotes/DAVe3283/dev' into change-i2c-freq

DAVe3283 committed:
Merge remote-tracking branch 'remotes/upstream/dev' into change-i2c-freq

DAVe3283 committed:
Merge remote-tracking branch 'remotes/upstream/dev' into change-i2c-freq

DAVe3283 commented:
I've been running this on ESPHome 2025.9.0b1 with the main I²C bus at 1 MHz and a virtual bus off the tca9548a at 300 kHz for over a week with no issues at all.

<img width="766" height="401" alt="image" src="https://github.com/user-attachments/assets/4fea1591-beec-43cf-b99b-49fc21d3a597" />

DAVe3283 committed:
Merge remote-tracking branch 'remotes/upstream/dev' into change-i2c-freq

DAVe3283 committed:
Merge remote-tracking branch 'remotes/upstream/dev' into change-i2c-freq

jesserockz commented on esphome/components/i2c/i2c_bus_arduino.cpp:
```diff
@@ -65,9 +65,11 @@ void ArduinoI2CBus::set_pins_and_clock_() {
+  this->set_clock_();
 }
 
+void ArduinoI2CBus::set_clock_() { wire_->setClock(frequency_); }
```
```suggestion
void ArduinoI2CBus::set_clock_() { this->wire_->setClock(this->frequency_); }
```

jesserockz commented on esphome/components/i2c/i2c_bus_esp_idf.cpp:
```diff
@@ -77,6 +69,29 @@ void IDFI2CBus::setup() {
   }
 }
 
+ErrorCode IDFI2CBus::set_clock_() {
```
`set_clock_` doesn't match the body of this function and could be confusing

jesserockz commented on esphome/components/i2c/i2c_bus_esp_idf.cpp:
```diff
@@ -184,6 +199,14 @@ ErrorCode IDFI2CBus::write_readv(uint8_t address, const uint8_t *write_buffer, s
 }
 
+ErrorCode IDFI2CBus::set_frequency(uint32_t frequency) {
+  frequency_ = frequency;
```
```suggestion
  this->frequency_ = frequency;
```

swoboda1337 commented on esphome/components/i2c/i2c_bus_esp_idf.cpp:
```diff
@@ -77,6 +69,29 @@ void IDFI2CBus::setup() {
+  }
+  return ERROR_OK;
+}
+
```
`set_clock _` is not a great name as this creates the device config, does why more than set the clock, maybe something like update_device?

swoboda1337 replied:
I also don't really like added and removing devices configs. However I understand why it was done this way. Not sure if there is a better way to do this without a major api change.

DAVe3283 replied:
I agree it is not ideal, but that is the only relevant ESP-IDF API I found. Every other solution I came up with was far more hacky than this :(

swoboda1337 replied:
I will think about this but yeah I understand why you did it this way.

swoboda1337 commented on esphome/components/i2c/i2c_bus_arduino.cpp:
```diff
@@ -65,9 +65,11 @@ void ArduinoI2CBus::set_pins_and_clock_() {
+  this->set_clock_();
 }
 
+void ArduinoI2CBus::set_clock_() { wire_->setClock(frequency_); }
```
Should just remove this function and do `wire_->setClock(frequency_);` instead.

DAVe3283 replied:
Done.

swoboda1337 commented on esphome/components/tca9548a/tca9548a.cpp:
```diff
@@ -29,18 +33,31 @@ void TCA9548AComponent::dump_config() {
-  uint8_t channel_val = 1 << channel;
+  if (frequency) {
+    this->original_frequency_ = this->bus_->get_frequency();
+    const i2c::ErrorCode err = this->bus_->set_frequency(frequency);
```
Should check to see if current frequency is the same as the requested one and avoid the frequency change.

DAVe3283 replied:
Good call. I now handle this in the I²C driver itself (Arduino or ESP-IDF), rather than trust all components to catch this and do it right.

swoboda1337 commented on esphome/components/tca9548a/tca9548a.cpp:
```diff
@@ -8,13 +8,17 @@ static const char *const TAG = "tca9548a";
 i2c::ErrorCode TCA9548AChannel::write_readv(uint8_t address, const uint8_t *write_buffer, size_t write_count,
                                             uint8_t *read_buffer, size_t read_count) {
-  auto err = this->parent_->switch_to_channel(channel_);
+  auto err = this->parent_->switch_to_channel(channel_, this->frequency_);
```
Can't the call to `this->parent_->bus_->set_frequency(frequency);` go here instead of passing the frequency to the parent?

DAVe3283 replied:
It could. My thought was that combining the channel and frequency switch into one function would reduce the chances of selecting a channel without also selecting the user's specified frequency. But if it reads cleaner integrating it here, that is probably the better way. I'll update the PR after dinner.

DAVe3283 committed:
Tweaks based on PR feedback

- Don't change frequency when new frequency is the same as current frequency
- Apply frequency changes in the TCA9548A *channel*, not the *component*

jesserockz commented on esphome/components/i2c/i2c_bus_arduino.cpp:
```diff
@@ -163,6 +163,17 @@ ErrorCode ArduinoI2CBus::write_readv(uint8_t address, const uint8_t *write_buffe
+  }
+  this->frequency_ = frequency;
+  if (this->initialized_) {
+    wire_->setClock(frequency_);
```
> All uses of class members and member functions should be prefixed with this-> to distinguish them from global functions/variables.

https://developers.esphome.io/contributing/code/#c-code-style

```suggestion
    this->wire_->setClock(this->frequency_);
```

DAVe3283 replied:
Fixed, and fixed several other places it was missed in the files I am already touching for this PR.

DAVe3283 committed:
Comply with style guide 🧹

bdraco committed:
Merge branch 'dev' into change-i2c-freq

DAVe3283 committed:
Merge remote-tracking branch 'remotes/upstream/dev' into change-i2c-freq

DAVe3283 commented:
Been about a month since flashing on ESPHome 2025.10.2. Perfectly stable and no memory leak:

<img width="763" height="392" alt="image" src="https://github.com/user-attachments/assets/d280eb25-a8a4-48f7-bf21-2a6a51c55909" />

DAVe3283 committed:
Support new I²C bus

DAVe3283 committed:
Merge remote-tracking branch 'remotes/upstream/dev' into change-i2c-freq

DAVe3283 committed:
Use LOG_STR with status_set_error

DAVe3283 committed:
Merge remote-tracking branch 'remotes/upstream/dev' into change-i2c-freq

DAVe3283 committed:
Merge remote-tracking branch 'remotes/upstream/dev' into change-i2c-freq

DAVe3283 committed:
Implement variable I²C frequency for HostI2CBus
